### ISSUE TITLE
Proposal: Improved error messages and handling

### ISSUE THREAD

**Issue by matthewp** [donejs/donejs#1073]
__tldr;__ put some effort into improving our error messaging across the board in DoneJS, but especially in StealJS and done-ssr.

> This was discussed on a [recent live stream (30:19)](https://youtu.be/ndAlLvRmyNI?t=30m19s).

# The Problem

*live-reload* is a fantastic tool for helping with rapid development of application. But if your app gets into an awkward state, like when you have a syntax error, it can be difficult to understand what the problem is and figure out how to fix it.

# The Solution

We would like to put some effort into improving our error messaging. Some inspiration comes from other frameworks, such as create-react-app which gives errors like:

<img width="536" alt="687474703a2f2f692e696d6775722e636f6d2f466e4c566677362e706e67" src="https://user-images.githubusercontent.com/361671/35401486-a1d565de-01c7-11e8-9441-1ea636b9cf73.png">

When there is a syntax error.

Elm gives good error messages when there is a compilation error:

![incomplete](https://user-images.githubusercontent.com/361671/35401758-66da1460-01c8-11e8-9cb1-3836b190c75e.png)

What if we could produce error messages that specifically showed bad code. Imagine code like:

```js
let map = new DefineMap();

map.set('bar');
```

What if we could display an error message (in the console? Maybe in the browser some how too?) that was:

```
Missing required argument.

2|> let map = new DefineMap();
3|>
4|> map.set('bar');

`set` expects to be passed 2 arguments.
```

This would be quite useful as it would give a *dynamic* language like JavaScript some of (but not all of course) the benefits of static languages by pointing out mistakes.

# Tasks

We can't just jump to implementation as we really need to get a good handle on what users are actually facing. I think the steps would be:

1. Spend a few days just trying to break DoneJS in as many ways as possible, documenting what was done and what the result is (a screenshot if possible).
2. Spend a few more days doing some analysis of other tools and how they handle these types of mistakes.
3. Build a system so that we can gather more detailed information on errors.
4. Design how we will display these errors. Should it be done in the console? Or specially handled in each environment (maybe an overlay in the browser and special formatting in the terminal?). Research in __(2)__ will inform our choices here.
5. Implement and show off!

**Comment by justinbmeyer** [comment]
Could we switch to `development.html` automatically, or provide a link if there's an error?

**Comment by matthewp** [comment]
## Errors

| Scenario    | File missing                                                                                                                           | Syntax error |
| ----------- | -------------------------------------------------------------------------------------------------------------------------------------- | ------------ |
| Loading     | https://github.com/stealjs/steal/issues/1331 https://github.com/stealjs/steal/issues/1332 https://github.com/stealjs/steal/issues/1333 | https://github.com/stealjs/steal/issues/1335 |
| SSR         | https://github.com/donejs/done-serve/issues/98 https://github.com/donejs/done-serve/issues/99                                          |              |
| Live Reload | Causes browser refresh (is that ok?)                                                                                                   |              |

**Comment by justinbmeyer** [comment]
Another stealjs one that I’m not sure is already reported, what happens if package.json isn’t valid, is missing a main, etc.

**Comment by matthewp** [comment]
Thanks, i'll do that one as well.

**Comment by matthewp** [comment]
As far as architecting error messaging, this is what I'm currently thinking (mostly steal, but application elsewhere in donejs as well):

* Error detection can happen any where in the codebase.
  * Where errors are detected, they should *not* be logged (warnings should though).
  * Instead they should be passed up the chain using `Promise.reject(err)`.
* Errors should have clear `message` properties.
* We want to have custom stack traces that remove the irrelevant steal lines when we *know* where the error occurs.
  * This _might_ mean subclassing `Error`.
* Errors are reported (logged) in steal's startup().

**Comment by matthewp** [comment]
Errors during done-ssr should:

* [ ] - Forward to the client.
* [ ] - Only log in done-serve, and only in production mode.
* During a reload it should:
  * [ ] - Forward the error to the client for each request once an error occurs.
  * [ ] - Work once the error is fixed.

**Comment by matthewp** [comment]
| Scenario                                                 | Server response   | Problems                                                                                                                               | Issues |
|----------------------------------------------------------|-------------------|----------------------------------------------------------------------------------------------------------------------------------------|--------|
| Error starting done-serve                                | The error message |                                                                                                                                        |        |
| Error starting done-serve, then it is fixed.             | Should render     | Fixing itself is hard because file-watching is based on getting the graph of the project, which we cannot do if there is a load error. |        |
| done-serve starts, then an error occurs on a file change | The error message |                                                                                                                                        |        |
| done-serve starts, then an error, then it is fixed       | Should render     |                                                                                                                                        |        |
